packages/teleterm/src/ui/DocumentCluster/ClusterResources/Kubes/useKubes.ts

```diff
@@ -14,18 +14,63 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
+import { AttemptStatus } from 'shared/hooks/useAsync';
+
+import { Kube, ServerSideParams } from 'teleterm/services/tshd/types';
+import { useAppContext } from 'teleterm/ui/appContextProvider';
 import { useClusterContext } from 'teleterm/ui/DocumentCluster/clusterContext';
 
+import { useServerSideResources } from '../useServerSideResources';
+
+function getEmptyTableText(status: AttemptStatus) {
+  switch (status) {
+    case 'error':
+      return 'Failed to fetch kubernetes clusters.';
+    case '':
+      return 'Searching…';
+    case 'processing':
+      return 'Searching…';
+    case 'success':
+      return 'No kubernetes clusters found.';
+  }
+}
+
 export function useKubes() {
+  const appContext = useAppContext();
   const ctx = useClusterContext();
-  const kubes = ctx.getKubes();
-  const syncStatus = ctx.getSyncStatus().kubes;
+  const {
+    fetchAttempt,
+    updateQuery,
+    agentFilter,
+    prevPage,
+    nextPage,
+    updateSearch,
+    onAgentLabelClick,
+    updateSort,
+    pageCount,
+  } = useServerSideResources<Kube>(
+    { fieldName: 'name', dir: 'ASC' }, // default sort
+    (params: ServerSideParams) => appContext.resourcesService.fetchKubes(params)
+  );
```

Comment:
We do a lot of destructuring only to return most of those values from `useKubes` without touching them.

I experimented with this a bit and arrived at this:

<details>

```diff
diff --git a/packages/teleterm/src/ui/DocumentCluster/ClusterResources/Kubes/useKubes.ts b/packages/teleterm/src/ui/DocumentCluster/ClusterResources/Kubes/useKubes.ts
index 77ef9469..a8dd125d 100644
--- a/packages/teleterm/src/ui/DocumentCluster/ClusterResources/Kubes/useKubes.ts
+++ b/packages/teleterm/src/ui/DocumentCluster/ClusterResources/Kubes/useKubes.ts
@@ -38,33 +38,17 @@ function getEmptyTableText(status: AttemptStatus) {
 export function useKubes() {
   const appContext = useAppContext();
   const ctx = useClusterContext();
-  const {
-    fetchAttempt,
-    updateQuery,
-    agentFilter,
-    prevPage,
-    nextPage,
-    updateSearch,
-    onAgentLabelClick,
-    updateSort,
-    pageCount,
-  } = useServerSideResources<Kube>(
+  const { updateSort, ...serverSideResources } = useServerSideResources<Kube>(
     { fieldName: 'name', dir: 'ASC' }, // default sort
     (params: ServerSideParams) => appContext.resourcesService.fetchKubes(params)
   );
+  const { fetchAttempt, agentFilter } = serverSideResources;
 
   return {
+    ...serverSideResources,
     connect: ctx.connectKube,
-    fetchAttempt,
     kubes: fetchAttempt.data?.agentsList,
-    agentFilter,
-    updateQuery,
-    updateSearch,
-    onAgentLabelClick,
-    pageCount,
     disabledRows: fetchAttempt.status === 'processing',
-    nextPage,
-    prevPage,
     emptyTableText: getEmptyTableText(fetchAttempt.status),
     customSort: {
       dir: agentFilter.sort?.dir,
```
</details>

This will become even cleaner if we refactor `customSort` so that it's returned by `useServerSideResources` instead, then we wouldn't have to make special accommodations for it. From what I see, `customSort` is being constructed in the same way in all three hooks.

I checked and it doesn't impair Typescript's "Go to definition" feature. You can go to definition of `updateQuery` in Kubes.tsx and it takes you straight to `useServerSideResources`.